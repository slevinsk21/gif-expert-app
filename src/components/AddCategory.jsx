// react
import { useState } from 'react';
import { func } from 'prop-types';


export const AddCategory = ({ onAddCategory }) => {
  const [value, setValue] = useState('');

  const handleSubmit = event => {
    event.preventDefault();
    value.trim().length && onAddCategory(value);
    setValue('');
  };

  return (
    <form
      aria-label='form'
      onSubmit={handleSubmit}
      style={{ display: 'flex', margin: '8px 0' }}
    >
      <input
        type='text'
        placeholder='Buscar Gifs...'
        value={value}
        onChange={({ target }) => setValue(target.value)}
        maxLength={32}
      />
    </form>
  );
};

AddCategory.propTypes = {
  onAddCategory: func.isRequired
};
