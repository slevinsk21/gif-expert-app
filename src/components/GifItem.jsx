// react
import { string } from 'prop-types';

export const GifItem = ({ title, url }) => {
  return (
    <div className='card' style={{ fontSize: 14 }}>
      <img src={url} alt={`${title}-image`} />

      <p>{title}</p>
    </div>
  );
};

GifItem.propTypes = {
  title: string.isRequired,
  url: string.isRequired
};
