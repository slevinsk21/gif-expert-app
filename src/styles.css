/* globals */
* {
  font-family: 'Ubuntu', sans-serif;
}
html {
  background-color: #252525;
  color: #f3f3f3;
  padding: 12px;
}
input {
  background-color: white;
  border-radius: 5px;
  border: 1px solid rgb(97, 32, 158);
  color: black;
  font-size: 1.2rem;
  outline: none;
  padding: 10px 15px;
  width: 100%;
}
h2 {
  font-size: 1.5rem;
}
h3 {
  font-size: 3rem;
  margin-bottom: 5px;
}

/* card */
.card-grid {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
}
.card {
  align-content: center;
  align-items: center;
  background-color: white;
  border-radius: 10px;
  box-shadow: 0px 3px 5px rgba(0, 0, 0, 0.03);
  display: flex;
  flex-direction: column;
  height: 0%;
  justify-content: center;
  margin-bottom: 20px;
  margin-right: 20px;
  overflow: hidden;
  max-width: 220px;
}
.card p {
  background-color: white;
  color: #333;
  flex: 1;
  margin-top: 5px;
  padding: 5px 20px 0px 20px;
  text-align: center;
  text-transform: capitalize;
}
.card img {
  width: 100%;
}

/* fade */
.fade-in {
  transition: all 0.6s;
  animation: fadeIn 0.6s;
  -webkit-animation: fadeIn 0.6s;
  -moz-animation: fadeIn 0.6s;
  -o-animation: fadeIn 0.6s;
  -ms-animation: fadeIn 0.6s;
}

@keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

@-moz-keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

@-webkit-keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

@-o-keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

@-ms-keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
