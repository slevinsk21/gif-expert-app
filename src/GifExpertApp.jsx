// react
import { useState } from 'react';

// components
import { AddCategory, GifGrid } from './components';

export const GifExpertApp = () => {
  const [categories, setCategories] = useState(['One Punch', 'DragÃ³n ball']);

  const onAddCategory = newCategory => {
    const newValueExist = categories.find(
      i => i.toLowerCase() === newCategory.toLowerCase()
    );
    // const newValueExist = categories.includes(newCategory);
    !newValueExist
      ? setCategories([newCategory, ...categories])
      : alert(`la cattegorÃ­a: ${newCategory} ya fue agregada...`);
  };

  const onDeleteCategory = categoryToDelete => {
    setCategories(categories.filter(category => category !== categoryToDelete));
  };

  return (
    <>
      <h1>GifExpertApp</h1>

      <AddCategory onAddCategory={onAddCategory} />

      {categories.length ? (
        categories.map(category => (
          <GifGrid
            key={category}
            category={category}
            onDeleteCategory={onDeleteCategory}
          />
        ))
      ) : (
        <h1
          className='fade-in'
          style={{
            display: 'flex',
            justifyContent: 'center',
            maxWidth: 380,
            lineHeight: '42px',
            margin: '50px auto',
            textAlign: 'center'
          }}
        >
          AÃºn no hay nada por aquÃ­ ðŸ˜…... por favor agrega un nombre
        </h1>
      )}
    </>
  );
};
