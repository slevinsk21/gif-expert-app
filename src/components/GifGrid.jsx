// react
import { string } from 'prop-types';

// custom hooks
import { useFetchGifs } from '../hooks';

// components
import { GifItem } from './GifItem';

export const GifGrid = ({ category, onDeleteCategory }) => {
  const { gifs, isLoading } = useFetchGifs(category);

  return (
    <div
      className='fade-in'
      style={{
        background: '#f4f4',
        padding: 10,
        borderRadius: 4,
        margin: '2px 0'
      }}
    >
      <div style={{ display: 'flex', justifyContent: 'space-between' }}>
        <h4>{category}</h4>

        {isLoading && <small>cargando...</small>}

        <button
          onClick={() => onDeleteCategory(category)}
          style={{
            borderRadius: 100,
            padding: '2px 6px',
            boxSizing: 'border-box',
            backgroundColor: '#252525',
            border: '1px solid black',
            alignSelf: 'flex-start',
            cursor: 'pointer'
          }}
        >
          x
        </button>
      </div>

      {!isLoading && gifs.length && (
        <div className='card-grid'>
          {gifs.map(image => (
            <GifItem key={image.id} {...image} />
          ))}
        </div>
      )}
    </div>
  );
};

GifGrid.propTypes = {
  category: string.isRequired
}