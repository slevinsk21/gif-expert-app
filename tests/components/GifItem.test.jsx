import { render, screen } from '@testing-library/react';

import { GifItem } from '../../src/components';

describe('File: <GifItem>', () => {
  const title = 'One Punch';
  const url = 'https://media2.giphy.com/media/l3vRlt8kty8KKeHni/giphy.gif?cid=31de32d5k2muizam1hx9t4pnszuerbpp1gl0rj6nx6wxry0k&rid=giphy.gif&ct=g';

  test('#1: Test: ( must be do match to the snapshot )', () => {
    const { container } = render(<GifItem title={title} url={url} />);

    expect(container).toMatchSnapshot();
  });

  test('#2 Test: ( should be show the image with the url and alt properly indicated )', () => {
    render(<GifItem title={title} url={url} />);
    const { src, alt } = screen.getByRole('img');

    expect(src).toBe(url);
    expect(alt).toBe(`${title}-image`);
  });

  test('#3 Test: ( should be show the title on the component )', () => {
    render(<GifItem title={title} url={url} />);

    expect(screen.getByText(title));
  });
});
