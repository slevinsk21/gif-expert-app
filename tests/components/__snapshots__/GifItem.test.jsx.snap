// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File: <GifItem> #1: Test: ( must be do match to the snapshot ) 1`] = `
<div>
  <div
    class="card"
    style="font-size: 14px;"
  >
    <img
      alt="One Punch-image"
      src="https://media2.giphy.com/media/l3vRlt8kty8KKeHni/giphy.gif?cid=31de32d5k2muizam1hx9t4pnszuerbpp1gl0rj6nx6wxry0k&rid=giphy.gif&ct=g"
    />
    <p>
      One Punch
    </p>
  </div>
</div>
`;
